Linear Tree Implementation

bammari opened the pull request:
This PR now allows users to embed linear model decision trees (trained with the linear-tree package https://github.com/cerlymarco/linear-tree) within OMLT. We include several univariate and bivariate tests to ensure the several implemented formulations are functional. Furthermore, this PR includes a preliminary Jupyter notebook that shows how to embed these linear model decision trees. using OMLT

In addition to this implementation, this PR also rewrites the way scaled_input_bounds are defined within the _setup_scaled_inputs_and_outputs function in formulation.py. Specifically, we pass a dictionary of tuples containing lower and upper bounds rather than constructing a bounds rule. test_maxpool_FullSpaceNNFormulation was failing without this change.

**Legal Acknowledgement**\
By contributing to this software project, I agree my contributions are submitted under the BSD license. 
I represent I am authorized to make the contributions and grant the license. 
If my employer has rights to intellectual property that includes these contributions, 
I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

carldlaird commented on .github/workflows/main.yml:
```diff
@@ -35,6 +35,7 @@ jobs:
           python -m pip install --upgrade pip setuptools wheel
           python -m pip install --upgrade coverage[toml] virtualenv tox tox-gh-actions          
           conda install -c conda-forge ipopt
+          conda install -c conda-forge pyscipopt
```
This is interesting. Can we use SCIP for anything else?

carldlaird commented on docs/notebooks/linear_trees/linear_tree_formulations.ipynb:
Please double-check this.

bammari replied:
I double checked the notebook. I added a quick section about saving trained linear model decision trees using pickle.

carldlaird commented on src/omlt/linear_tree/lt_definition.py:
```diff
@@ -0,0 +1,408 @@
+        Raises:
+            Exception: Input bounds required. If unscaled_input_bounds and
+                scaled_input_bounds is None, raise Exception.
+        """
```
It looks like a different doc-string format than I am used to. Can you verify that these work in the API docs?

bammari replied:
I generated the API docs and this format is fine. Note that I also added an omlt.linear_tree.rst file in docs\api_docs\ to generate the API docs using tox.

carldlaird commented on src/omlt/linear_tree/lt_definition.py:
```diff
@@ -0,0 +1,408 @@
+                raise ValueError(
+                    "Input Bounds needed to represent linear trees as MIPs"
+                )
+
```
It is unfortunate that we need to copy this from network_definition. A sign that we don't yet have the best abstraction. Let's consider this during a future refactoring.

bammari replied:
Agreed

carldlaird commented on src/omlt/linear_tree/lt_definition.py:
```diff
@@ -0,0 +1,408 @@
+
+    # Same as above but with right child
+    right_child = splits_dict[split]["children"][1]
+    if right_child in splits_dict:
```
The lines below were not covered. Seems odd. Do we need a different test?

bammari replied:
Never ran due to the structure of the trained tree. I added another test that ensures these children finders work regardless of tree structure.

carldlaird commented on src/omlt/linear_tree/lt_definition.py:
```diff
@@ -0,0 +1,408 @@
+        count = 0
+        # Checks to ensure that the input nested dictionary contains the
+        # correct information
+        for entry in model:
```
Something is strange in this code block below. (Maybe an indentation problem.). Please double-check.

bammari replied:
You are correct. I consolidated this into one else statement.

carldlaird commented on tests/linear_tree/test_lt_formulation.py:
```diff
@@ -0,0 +1,728 @@
+    lt_def2 = LinearTreeDefinition(
+        regr, unscaled_input_bounds=unscaled_input_bounds, scaling_object=scaler
+    )
+    assert lt_def2.scaled_input_bounds[0][0] - scaled_input_bounds[0][0] <= 1e-5
```
Why aren't these checks absolute value? Or maybe use pytest.approx?

carldlaird replied:
Maybe there are others like this as well.

bammari replied:
Thanks for noticing this. I changed all of them to use pytest.approx.

carldlaird commented on tests/linear_tree/test_lt_formulation.py:
```diff
@@ -0,0 +1,728 @@
+    y_pred = regr.predict(
+        np.array([pe.value(model1.x0), pe.value(model1.x1)]).reshape(1, -1)
+    )
+    assert y_pred[0] - solution_1_bigm <= 1e-4
```
Like above. Use Pytest.approx. (Check for others).

bammari replied:
Changed all to use pytest.approx!

rmisener commented on docs/notebooks/linear_trees/linear_tree_formulations.ipynb:
Would it be possible to arrange the placement of the notebooks so that [bo_with_trees.ipynb](https://github.com/cog-imperial/OMLT/blob/main/docs/notebooks/bo_with_trees.ipynb) is in the same folder as this new notebook? At least it's confusing to me why the notebooks for the trees are in separate places. If linear trees are so different from the GBT implementations that they do need to be in a different folder, perhaps add one more folder so that it's notebooks/trees and then have everything inside?

bammari replied:
Per your suggestion, I consolidated both of these notebooks in notebooks/trees.

In addition, I updated docs/notebooks.rst.

rmisener reviewed (changes requested):
Just noticed something small while working through some material, this should be easy.

rmisener commented:
Going through minor things and want to make sure @bammari gets credit for this work. Would you please also add to this PR:

In README.rst

- Please add a reference to the Computers & Chemical Engineering paper. State that if anyone uses the linear tree implementation, they should cite the C&CE paper.
- Please add @bammari to the list of contributors at the bottom and state his funding.

In the new notebook, please add the C&CE paper as a reference. As an example of how to do this, see bo_with_trees.ipynb

bammari commented:
> Going through minor things and want to make sure @bammari gets credit for this work. Would you please also add to this PR:
> 
> In README.rst
> 
> * Please add a reference to the Computers & Chemical Engineering paper. State that if anyone uses the linear tree implementation, they should cite the C&CE paper.
> * Please add @bammari to the list of contributors at the bottom and state his funding.
> 
> In the new notebook, please add the C&CE paper as a reference. As an example of how to do this, see bo_with_trees.ipynb

Added reference to CACE paper in both the notebook and Readme.rst. I also updated the OMLT paper reference from Arxiv to Journal of ML Research in README.rst. In addition, added contribution statement - Bashar
